src/utils.tsx

```diff
@@ -17,17 +31,17 @@ export function renderElement(renderer, props: any) {
 
 export type GetProps<C> = C extends React.ElementType<infer P> ? P : never;
 
-export function normalizeColumns<T extends React.ElementType = React.ElementType<any>>(elements: React.ElementType<T>) {
-  type P = GetProps<T>;
-  const columns: P[] = [];
-  React.Children.forEach(elements, element => {
+export function normalizeColumns(elements: React.ReactElement<IColumnProps>[]) {
+  const columns: IColumnProps[] = [];
+  React.Children.forEach<React.ReactElement<IColumnProps>>(elements, element => {
     if (React.isValidElement(element) && element.key) {
-      const column = { ...element.props, key: element.key };
-      columns.push(column as P);
+      const column: IColumnProps = { ...element.props, key: element.key };
+      columns.push(column);
     }
   });
   return columns;
 }
+export type TColumns = ReturnType<typeof normalizeColumns>;
```

Comment:
defining a type based on what the function returns is very prone to bugs. For example, you could add another conditional branch and return something completely unexpected and typescript would not be able to help you.

the idea is for the function to actually match what is expected for whatever is using it